### PR TITLE
`rb-deep-search` component

### PR DESCRIPTION
**Pull request by jonman**
TP: https://rockabox.tpondemand.com/entity/9656

### CODE REVIEW

File: src/rb-deep-search/rb-deep-search-node.js [hunk 1]
```diff
@@ -0,0 +1,113 @@
+define([
+], function () {
+
+    function Node (node, parent) {
+        this._children = [];
+        this._matched = false;
+        this._node = node;
+        this._parent = parent;
+        this._partial = false;
```
Comment (jonman) [inline src/rb-deep-search/rb-deep-search-node.js]: I'm thinking `partial` could be renamed too `indeterminate`?

File: src/rb-deep-search/demo/demo-ctrl.js [hunk 2]
```diff
@@ -0,0 +1,64 @@
+define([
+], function () {
+
+    // @ngInject
+    function demoCtrl () {
+
+        this.categories = [
+            {
+                anchor: 'Pokedex',
+                label: 'Pokedex',
```
Comment (nwhite89) [inline src/rb-deep-search/demo/demo-ctrl.js]: :+1:

File: test/unit/rb-deep-search/rb-deep-search.spec.js [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+                $scope.categories = categories;
+                $scope.ngModel = ngModel;
+                element = $compile(template)($scope);
+                $scope.$apply();
+                isolatedScope = element.isolateScope();
+            };
+        }));
+
+        describe('', function () {
+            it('should ', function () {
```
Comment (nwhite89) [inline test/unit/rb-deep-search/rb-deep-search.spec.js]: ?

Comment (jonman) [inline test/unit/rb-deep-search/rb-deep-search.spec.js]: Loading tests ... please wait ...

Comment (nwhite89) [inline test/unit/rb-deep-search/rb-deep-search.spec.js]: haha ok thought that was the case as it was DNM just thought would check :+1:

File: src/rb-deep-search/rb-deep-search-node.js [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        this._node = node;
+        this._parent = parent;
+
+        this.items = [];
+        this.label = node.label || '';
+
+        this.init();
+    }
+
+    Node.prototype.each = function (collection, iteratee) {
```
Comment (nwhite89) [inline src/rb-deep-search/rb-deep-search-node.js]: could you add some JSDoc information please

File: src/rb-deep-search/rb-deep-search-tree.js [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+], function (Node) {
+
+    function Tree (nodes) {
+        this._nodes = nodes;
+        this.items = [];
+
+        this.init();
+    }
+
+    Tree.prototype.each = function (collection, iteratee) {
```
Comment (nwhite89) [inline src/rb-deep-search/rb-deep-search-tree.js]: Same as previous with docs, possibly split the each function out as it's a duplicate?

File: CHANGELOG.md [hunk 6]
```diff
@@ -19,6 +19,8 @@ This project adheres to [Semantic Versioning](http://semver.org/).
 - [`rb-page-title` `compact` modifier](https://github.com/rockabox/rbx_ui_components/pull/230)
 - [`rb-side-nav` `save-button-disabled` attribute allows for disabling the save button](https://github.com/rockabox/rbx_ui_components/pull/237)
 - [`rb-side-nav` `valid` and `invalid` will display a status badge of a tick or ! when true](https://github.com/rockabox/rbx_ui_components/pull/231)
+- [`rb-deep-search` component](https://github.com/rockabox/rbx_ui_components/pull/232)
+- ['rb-scrollspy' do not show hidden sections](https://github.com/rockabox/rbx_ui_components/pull/232)
```
Comment (chrisvxd) [inline CHANGELOG.md]: Is this meant to be part of this PR?

**Comment by chrisvxd** [comment]
Added squash before merge label

File: src/rb-deep-search/demo/demo-categories-pokedex.js [hunk 7]
```diff
@@ -0,0 +1,59 @@
+define([
+], function () {
+
+    return [
+        {
+            anchor: 'Pokedex',
```
Comment (chrisvxd) [inline src/rb-deep-search/demo/demo-categories-pokedex.js]: A fine choice of test data

File: src/rb-deep-search/rb-deep-search-node.js (hunk 4 above)
Comment (chrisvxd) [inline src/rb-deep-search/rb-deep-search-node.js]: +1

File: src/rb-deep-search/rb-deep-search-node.js [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+        if (!this._protected && this.label && this.label.toLowerCase().indexOf(query) > -1) {
+            this.hidden = false;
+        } else {
+            this._node.hidden = true;
+            // Keep searching through child nodes
+            this.each(this.items, function (item) {
+                item.search(query);
+            });
+        }
+    };
```
Comment (chrisvxd) [inline src/rb-deep-search/rb-deep-search-node.js]: Nice recursive approach!

File: src/rb-deep-search/rb-deep-search-tree.js [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+
+    Tree.prototype.search = function (query) {
+        this.reset();
+        this.each(this.items, function (item) {
+            item.search(query);
+        });
+    };
+
+    return Tree;
+
```
Comment (chrisvxd) [inline src/rb-deep-search/rb-deep-search-tree.js]: Why do we have both `Tree` and `Node?

File: src/rb-scrollspy/demo/demo.tpl.html [hunk 10]
```diff
@@ -21,25 +21,25 @@
     </div>
 
     <div class="search">
-        <input type="text" ng-model="search" placeholder="Search list...">
+        <rb-deep-search
+            ng-model="demoCtrl.filteredCategories"
```
Comment (chrisvxd) [inline src/rb-scrollspy/demo/demo.tpl.html]: `filteredCategories` can probably be renamed now

File: CHANGELOG.md (hunk 6 above)
Comment (jonman) [inline CHANGELOG.md]: I had to plug the deep search into the scroll spy demo so it made sense to add that update in here?

Comment (chrisvxd) [inline CHANGELOG.md]: Noticed that, ignore me

File: src/rb-deep-search/rb-deep-search-tree.js (hunk 9 above)
Comment (jonman) [inline src/rb-deep-search/rb-deep-search-tree.js]: The parent structure is slightly different to inner parts. Parent is essentially an array of objects `[{},{},{}]` whereas the inner objects are iterated via `{items:[]}`.

**Pull request merged**